substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/AnalysisMethod.java

```diff
@@ -217,25 +219,28 @@ public void registerAsEntryPoint(Object newEntryPointData) {
         startTrackInvocations();
     }
 
-    public void registerAsInvoked(InvokeTypeFlow invoke) {
-        isInvoked = true;
+    public boolean registerAsInvoked(InvokeTypeFlow invoke) {
         if (invokedBy != null && invoke != null) {
             invokedBy.put(invoke, Boolean.TRUE);
         }
+        return atomicMark(isInvoked);
     }
 
-    public void registerAsImplementationInvoked(InvokeTypeFlow invoke) {
+    public boolean registerAsImplementationInvoked(InvokeTypeFlow invoke) {
         assert !Modifier.isAbstract(getModifiers());
-        isImplementationInvoked = true;
         if (implementationInvokedBy != null && invoke != null) {
             implementationInvokedBy.put(invoke, Boolean.TRUE);
         }
 
-        /*
-         * The class constant of the declaring class is used for exception metadata, so marking a
-         * method as invoked also makes the declaring class reachable.
-         */
-        getDeclaringClass().registerAsReachable();
+        if (atomicMark(isImplementationInvoked)) {
```

Comment:
This has a subtle side effect: `registerAsImplementationInvoked` can now return before the class is actually marked as reachable (because that can still be pending in another thread). I don't think it is worth risking potential concurrency bugs where this behavior is unexpected, so please leave the `registerAsReachable` before the `if` so that it is always invoked. And add a comment documenting that.

Reply:
Good catch. Done.